modules/library/coverage/src/main/java/org/geotools/image/util/ImageUtilities.java

```diff
@@ -1361,18 +1361,74 @@ public static void setSubsamplingFactors(
-        int subSamplingFactorX = (int) Math.round(requestedRes[0] / selectedRes[0]);
-        subSamplingFactorX = subSamplingFactorX == 0 ? 1 : subSamplingFactorX;
 
-        while (subSamplingFactorX > 0 && rasterWidth / subSamplingFactorX <= 0) subSamplingFactorX--;
-        subSamplingFactorX = subSamplingFactorX <= 0 ? 1 : subSamplingFactorX;
-
-        int subSamplingFactorY = (int) Math.round(requestedRes[1] / selectedRes[1]);
-        subSamplingFactorY = subSamplingFactorY == 0 ? 1 : subSamplingFactorY;
+        // Control parameters
+        final double resEpsilon = 1e-9; // Small epsilon to avoid floating point errors when using floor
+        final double overviewTolerance =
+                0.03; // Relative tolerance (%) to allow a small upward adjustment of subsampling
+
+        // Ratios between requested resolution and the resolution of the selected overview
+        final double ratioX = requestedRes[0] / selectedRes[0];
+        final double ratioY = requestedRes[1] / selectedRes[1];
+
+        // Use floor with epsilon and minimum clamps
+        int subSamplingFactorX = (int) Math.floor(ratioX + resEpsilon);
+        int subSamplingFactorY = (int) Math.floor(ratioY + resEpsilon);
+
+        subSamplingFactorX = Math.max(1, subSamplingFactorX);
+        subSamplingFactorY = Math.max(1, subSamplingFactorY);
+
+        // Explicit clamp per dimension (replaces integer division)
+        // Ensures we never request a factor larger than the raster size
+        if (subSamplingFactorX > rasterWidth) subSamplingFactorX = rasterWidth;
+        if (subSamplingFactorY > rasterHeight) subSamplingFactorY = rasterHeight;
+
+        // Upward adjustment ONLY if it falls within tolerance
+        // Prevents unnecessary quality drops, but allows better alignment when very close
+        int candidateX = subSamplingFactorX + 1;
+        if (candidateX <= rasterWidth) {
+            double relErrFloor = Math.abs(ratioX - subSamplingFactorX) / Math.max(ratioX, 1e-9);
+            double relErrUp = Math.abs(ratioX - candidateX) / Math.max(ratioX, 1e-9);
+            if (relErrUp <= overviewTolerance && relErrUp < relErrFloor) {
+                subSamplingFactorX = candidateX;
+            }
+        }
+        int candidateY = subSamplingFactorY + 1;
+        if (candidateY <= rasterHeight) {
+            double relErrFloor = Math.abs(ratioY - subSamplingFactorY) / Math.max(ratioY, 1e-9);
+            double relErrUp = Math.abs(ratioY - candidateY) / Math.max(ratioY, 1e-9);
+            if (relErrUp <= overviewTolerance && relErrUp < relErrFloor) {
+                subSamplingFactorY = candidateY;
+            }
+        }
 
-        while (subSamplingFactorY > 0 && rasterHeight / subSamplingFactorY <= 0) subSamplingFactorY--;
-        subSamplingFactorY = subSamplingFactorY <= 0 ? 1 : subSamplingFactorY;
+        // Align SourceRegion to subsampling factors to avoid seams
```

Comment:
Ah, this is new, interesting. (or did you pick it from some older code? I don't remember seeing region adjustment too).

The javadoc of the method should be updated accordingly.